cabal-install/Distribution/Client/SourceFiles.hs

```diff
@@ -0,0 +1,166 @@
+    else do
+        lhsExists <- doesFileExistMonitored "Setup.lhs"
+        when lhsExists $ need "Setup.lhs"
+
+needLibrary :: PackageDescription -> Library -> Rebuild ()
+needLibrary pkg_descr (Library { exposedModules = modules
+                               , signatures     = sigs
+                               , libBuildInfo   = bi })
+  = needBuildInfo pkg_descr bi (modules ++ sigs)
+
+needExecutable :: PackageDescription -> Executable -> Rebuild ()
+needExecutable pkg_descr (Executable { modulePath = mainPath
+                                     , buildInfo  = bi })
+  = do needBuildInfo pkg_descr bi []
+       needMainFile  bi mainPath
+
+needTestSuite :: PackageDescription -> TestSuite -> Rebuild ()
+needTestSuite pkg_descr t
+  = case testInterface t of
+      TestSuiteExeV10 _ mainPath -> do
+        needBuildInfo pkg_descr bi []
+        needMainFile  bi mainPath
+      TestSuiteLibV09 _ m ->
+        needBuildInfo pkg_descr bi [m]
+      TestSuiteUnsupported _ -> return () -- soft fail
+ where
+  bi = testBuildInfo t
+
+needMainFile :: BuildInfo -> FilePath -> Rebuild ()
+needMainFile bi mainPath = do
+    -- The matter here is subtle.  It might *seem* that we
+    -- should just search for mainPath, but as per
+    -- b61cb051f63ed5869b8f4a6af996ff7e833e4b39 'main-is'
+    -- will actually be the source file AFTER preprocessing,
+    -- whereas we need to get the file *prior* to preprocessing.
+    ppFile <- findFileWithExtensionMonitored
+                (ppSuffixes knownSuffixHandlers)
+                (hsSourceDirs bi)
+                (dropExtension mainPath)
+    case ppFile of
+        -- But check the original path in the end, because
+        -- maybe it's a non-preprocessed file with a non-traditional
+        -- extension.
+        Nothing -> findFileMonitored (hsSourceDirs bi) mainPath
+                    >>= maybe (return ()) need
+        Just pp -> need pp
+
+needBenchmark :: PackageDescription -> Benchmark -> Rebuild ()
+needBenchmark pkg_descr bm
+  = case benchmarkInterface bm of
+     BenchmarkExeV10 _ mainPath -> do
+       needBuildInfo pkg_descr bi []
+       needMainFile  bi mainPath
+     BenchmarkUnsupported _ -> return () -- soft fail
+ where
+  bi = benchmarkBuildInfo bm
+
+needBuildInfo :: PackageDescription -> BuildInfo -> [ModuleName] -> Rebuild ()
+needBuildInfo pkg_descr bi modules = do
+    -- NB: These are separate because there may be both A.hs and
+    -- A.hs-boot; need to track both.
+    findNeededModules ["hs", "lhs", "hsig", "lhsig"]
+    findNeededModules ["hs-boot", "lhs-boot"]
+    mapM_ needIfExists (cSources bi ++ jsSources bi)
+    -- A MASSIVE HACK to (1) make sure we rebuild when header
+    -- files change, but (2) not have to rebuild when anything
+    -- in extra-src-files changes (most of these won't affect
+    -- compilation).  It would be even better if we knew on a
+    -- per-component basis which headers would be used but that
+    -- seems to be too difficult.
```

Comment:
Yeah. Eventually we should make `extraSrcFiles` be per-component, and data files.